WalletWasabi/WabiSabi/Client/OwnershipProofValidator.cs

```diff
@@ -0,0 +1,127 @@
+
+			var mineTxIds = TransactionStore.GetTransactionHashes();
+			var alreadySeenCoins = othersCoins.Where(x => mineTxIds.Contains(x.OutPoint.Hash));
+
+			// In case one Alice is trying to spend an output from a transaction that we already have
+			// seen before then we can verify its script without downloading anything from the network.
+			foreach (var (_, outpoint, amount, ownershipProof) in alreadySeenCoins)
+			{
+				if (validProofs >= minimumNumberOfValidProofs)
+				{
+					break;
+				}
+
+				if (TransactionStore.TryGetTransaction(outpoint.Hash, out var stx))
+				{
+					var transactionCoins = stx.Transaction.Outputs.AsCoins();
+					var foundCoin = transactionCoins.ElementAtOrDefault((int)outpoint.N);
+
+					if (foundCoin is not { } coin || coin.Amount != amount)
+					{
+						throw new InvalidOperationException("The coordinator lies (coin doesn't exists or is different than the provided by the coordinator).");
+					}
+					VerifyCoin(foundCoin.ScriptPubKey, coinJoinInputCommitmentData, ownershipProof);
+					validProofs++;
+				}
+			}
+
+			// In case we cannot validate enough proofs using only our already seen transactions, we
+			// would need to start downloading the blocks containing the coins that need to be
+			// validated.
+			var coinsGroupedByBlocks = othersCoins.GroupBy(x => x.BlockId).ToArray();
+
+			// We sort the groups (blocks) randomly.
+			var random = new Random();
+			var randomNumbers = othersCoins.Select(r => random.Next()).ToArray();
+			coinsGroupedByBlocks = coinsGroupedByBlocks.Zip(randomNumbers, (group, rnd) => new { Groupd = group, Order = rnd })
+				.OrderBy(x => x.Order)
+				.Select(x => x.Groupd)
+				.ToArray();
+
+			foreach (var coinsInBlock in coinsGroupedByBlocks)
+			{
+				if (validProofs >= minimumNumberOfValidProofs)
+				{
+					break;
+				}
+
+				var blockId = coinsInBlock.Key;
+				var block = await BlockProvider.GetBlockAsync(blockId, cancellationToken).ConfigureAwait(false);
+
+				if (block is { })
+				{
+					foreach (var (_, outpoint, amount, ownershipProof) in coinsInBlock)
+					{
+						var tx = block.Transactions.FirstOrDefault(x => x.GetHash() == outpoint.Hash);
+						if (tx is null)
+						{
+							throw new InvalidOperationException($"The coordinator lies (there is not transaction with id '{outpoint.Hash}' in block '{blockId}').");
+						}
+						var transactionCoins = tx.Outputs.AsCoins();
+						var foundCoin = transactionCoins.ElementAtOrDefault((int)outpoint.N);
+						if (foundCoin is not { } coin || coin.Amount != amount)
+						{
+							throw new InvalidOperationException("The coordinator lies (coin doesn't exists or is different than the provided by the coordinator).");
+						}
+						VerifyCoin(foundCoin.ScriptPubKey, coinJoinInputCommitmentData, ownershipProof);
+						validProofs++;
+					}
+				}
+				else
+				{
+					// FIXME: what if no peer has the block?
```

Comment:
error

Reply:
What if the client has only one peer at that current time. Isn't that too restrictive to abort the CJ? We should only abort if we surely know that the coordinator lies. In any other cases, we could or retry.

Reply:
Hmm. Right now this is a trust issue, but this would be obviated if Wasabi kept track of block headers. I've been discussing this in slack with @lontivero.

I think the correct behavior would be to verify that all block hashes are part of the header chains, and then separately to spot check the txouts. An error in the spot checking can be tolerated but right now this branch might catch both failure paths, which should be handled differently.

In either case we'll have to get back to refining this, I slightly prefer erroring here for now but I don't mind logging instead.

Reply:
Add logging.